File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.java
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "null";

        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            TreeNode curr = q.poll();

            if (curr == null) {
                sb.append("null,");
                continue;
            }

            sb.append(curr.val).append(",");
            q.add(curr.left);
            q.add(curr.right);
        }

        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("null")) return null;

        String[] arr = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int i = 1;

        while (!q.isEmpty() && i < arr.length) {
            TreeNode curr = q.poll();

            if (!arr[i].equals("null")) {
                curr.left = new TreeNode(Integer.parseInt(arr[i]));
                q.add(curr.left);
            }
            i++;

            if (i < arr.length && !arr[i].equals("null")) {
                curr.right = new TreeNode(Integer.parseInt(arr[i]));
                q.add(curr.right);
            }
            i++;
        }

        return root;
    }
}

File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.cpp
class Codec {
public:
    string serialize(TreeNode* root) {
        string serStr = "";
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            TreeNode* front = q.front();
            q.pop();

            if (front == NULL) {
                serStr += "#,";
                continue;
            } else {
                serStr += to_string(front->val);
                serStr += ',';
            }
            q.push(front->left);
            q.push(front->right);
        }
        cout<<serStr<<endl;
        return serStr;
    }

    TreeNode* deserialize(string data) {
        if (data == "" || data == "#,") return nullptr;
        int ind=0;
        TreeNode* root=nullptr;
        vector<TreeNode*> treeVec;

        while(ind<data.length()){
            int nextComma=data.find(',',ind);
            string val=data.substr(ind,nextComma-ind);
            if(val=="#") treeVec.push_back(nullptr);
            else{
                TreeNode* newNode= new TreeNode(stoi(val));
                treeVec.push_back(newNode);
            }
            ind=nextComma+1;
        }
        int setInd=0,moveInd=1;

        while(moveInd<treeVec.size()){
            if(treeVec[setInd]==nullptr) {
                setInd++;
                continue;
            }
            treeVec[setInd]->left=treeVec[moveInd++];
            treeVec[setInd]->right=treeVec[moveInd++];
            setInd++;

        }
        return treeVec[0];
        
        return root;
    }
};